src/base/static/libs/maps/index.js

```diff
@@ -232,61 +244,54 @@ class MainMap {
   }
 
   geolocate() {
+    // TODO
     this.map.locate();
   }
 
-  addLayerView(collectionId) {
-    return model => {
-      this.layerViews[collectionId][model.cid] = new LayerView({
-        model: model,
-        router: this.router,
-        map: this.map,
-        layerGroup: this.layers[collectionId],
-        placeTypes: this.options.placeTypes,
-        // to access the filter
-        mapView: this,
-      });
-    };
-  }
-
-  removeLayerView(collectionId) {
-    return model => {
-      // remove map-bound events for this layer view
-      this.map.off(
-        "zoomend",
-        this.layerViews[collectionId][model.cid].updateLayer,
-        this.layerViews[collectionId][model.cid],
-      );
-      this.map.off(
-        "move",
-        this.layerViews[collectionId][model.cid].throttledRender,
-        this.layerViews[collectionId][model.cid],
-      );
-      this.map.off(
-        "zoomend",
-        this.layerViews[collectionId][model.cid].render,
-        this.layerViews[collectionId][model.cid],
-      );
-
-      this.layerViews[collectionId][model.cid].remove();
-      delete this.layerViews[collectionId][model.cid];
-
-      Util.log("APP", "panel-state", "closed");
-      if (this.locationTypeFilter) {
-        this.router.navigate("filter/" + this.locationTypeFilter, {
-          trigger: true,
-        });
-      } else {
-        this.router.navigate("/", { trigger: true });
-      }
-    };
+  updatePlaceCollectionLayer(collectionId, collection, modelId) {
+    this.map
+      .getSource(collectionId)
+      .setData(createGeoJSONFromCollection(collection, modelId));
   }
 
-  zoomInOn(/*latLng*/) {
-    // TODO
-    //this.map.setView(latLng, this.mapConfig.options.maxZoom || 17);
+  addPlaceCollectionLayer(collectionId, collection) {
+    this.layers[collectionId] = this.map.createGeoJSONLayer({
```

Comment:
I'm realizing that that POJO we are passing to `createGeoJSONLayer` is just a duplicate representation of our "layer config", which in my opinion, should be our first-class Layer representation. Thus, `addPlaceCollectionLayer` should accept a single argument - the first-class Layer (ie: the layer config).

That means the `place-collection:add-place` event which calls this method should pass the Layer as it's payload, which means that the emitter in InputForm should be passing the Layer instead of a `collectionId` and `collection`.

I think we can clarify the Layer abstraction with a helper function like `createLayerFromCollection` which accepts a `collection` and `collectionId` and returns a Layer. Then this method would look like this:

```
addPlaceCollectionLayer(layer) {
  this.layers[layer.id] = this.map.createGeoJSONLayer(layer)
  // etc
```

of course, I have my own suggestions about storing the first-class Layer representation in `this.layers` and combining `this.map.createXLayer` with `this.map.addLayer`, as described in my earlier comment.

What do you think?

Reply:
Ok, I think this makes sense. It was a little strange that the payload of the `place-collection:add-place` event was the Backbone collection itself.

So would `createLayerFromCollection()` live inside the Provider? If so, how should we make the Provider aware of our place collections?

> That means the place-collection:add-place event which calls this method should pass the Layer as it's payload

I think I'll have some additional questions down the road about how the `place-collection:add-place` should have access to the Layer representation in the first place. Should this `emitter` event actually be a Redux action?

Reply:
>Should this emitter event actually be a Redux action?
Let's not worry about Redux yet. It's more important that we establish a proper abstraction for our Layer first. I think the current emitter implementation is fine.

Reply:
So what I ended up doing here was to move the cache of map provider layers to an object called `layersCache` in the map provider module. In `MainMap`, `this._layers` is just a reference to the `layerConfig` with no transformations applied.

When we create a layer of any type, the map provider does three things:
* add a source for the layer's data
* build whatever internal layers it needs to faithfully produce the config layer, and cache this result in `layersCache`
* add the internal layers to the map

The `emitter` events for creating/updating place layers send along the Backbone collection id, then the collection itself is looked up in `MainMap` and we run the GeoJSON serialization routine. I didn't see a reason to send the first-class layer as the payload if the first-class representation lives in `MainMap`--but maybe I'm missing something there?

Hopefully this is a little cleaner-- let me know what you think.